apache2/waf_logging/waf_log_util.cc

```diff
@@ -54,12 +55,63 @@ static std::string to_json_string(const waf_logging::waf_log& log) {
     return json_output;
 };
 
-static std::vector<int> default_disableable_rule_ids;
+enum class rule_set_types {
+    TYPE_UNKNOWN,
+    OWASP_CRS,
+};
+
+enum class rule_set_versions {
+    VERSION_UNKNOWN,
+    VERSION_2_2_9,
+    VERSION_3_0,
+    VERSION_3_1,
+};
+
+using rule_set_info = std::pair<rule_set_types, rule_set_versions>;
+static const rule_set_info OWASP_2_2_9 {std::make_pair(rule_set_types::OWASP_CRS, rule_set_versions::VERSION_2_2_9)};
+static const rule_set_info OWASP_3_0 {std::make_pair(rule_set_types::OWASP_CRS, rule_set_versions::VERSION_3_0)};
+static const rule_set_info OWASP_3_1 {std::make_pair(rule_set_types::OWASP_CRS, rule_set_versions::VERSION_3_1)};
+
+static rule_set_types parse_rule_set_type(const char* type) {
+    if (!type) {
+        return rule_set_types::TYPE_UNKNOWN;
+    }
+
+    return (std::strcmp(type, "OWASP_CRS") == 0) ? rule_set_types::OWASP_CRS : rule_set_types::TYPE_UNKNOWN;
+}
+
+static rule_set_versions parse_rule_set_version(const char* version) {
+    if (!version) {
+        return rule_set_versions::VERSION_UNKNOWN;
+    }
+
+    return (std::strcmp(version, "2.2.9") == 0) ? rule_set_versions::VERSION_2_2_9 :
+        (std::strcmp(version, "3.0.0") == 0) ? rule_set_versions::VERSION_3_0 :
+        (std::strcmp(version, "3.1.0") == 0) ? rule_set_versions::VERSION_3_1 :
+        rule_set_versions::VERSION_UNKNOWN;
+}
+
+using rule_ids_vector = std::vector<int>;
+
+static rule_ids_vector default_disableable_rule_ids;
+static std::map<rule_set_info, rule_ids_vector> disableable_rule_ids_map;
 
 static const std::string WAF_ACTION_BLOCKED {"Blocked"};
 static const std::string WAF_ACTION_DETECTED {"Detected"};
 static const std::string WAF_ACTION_MATCHED {"Matched"};
 
+static bool is_mandatory_rule(const rule_set_types type, const rule_set_versions version, const int current_id) {
+    const auto is_disableable = [current_id] (const rule_ids_vector& rule_ids) {
```

Comment:
Can we not use the disableable term? We could come up with something more readable. is_valid perhaps?

Reply:
I think this particular helper lambda should rather be called `contains_current_id` because it checks the passed-in vector for the presence of the value. Let me rename it to better reflect what it does.

As for the more general term - agree it sounds funny but since the file is named so I have reflected it in some names in the code too. `is_valid` is not exactly that for a rule may be or not be in the file yet it still remains valid. This only tells us whether the rule is mandatory or can be disabled.